File: 139.cpp
struct Solution {
    bool wordBreak(const string &s, const vector<string>& wordDict) {
        vector<bool> dp(s.size() + 1, false);
        dp.back() = true;
        for (int i = s.size() - 1; i >= 0; --i){
            for (const string &w : wordDict){
                if (i + w.size() <= s.size() && s.substr(i, w.size()) == w && dp[i + w.size()]){
                    dp[i] = true;
                    break;
                }
            }
        }
        return dp[0];
    }
};

File: 139.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        dp = [False] * (len(s) + 1)
        dp[-1] = True
        for i in range(len(s) - 1, -1, -1):
            for w in wordDict:
                if i + len(w) <= len(s) and s[i:i + len(w)] == w and dp[i + len(w)]:
                    dp[i] = True
                    break
        return dp[0]

# class Solution:
#     def wordBreak(self, s: str, wordDict: List[str]) -> bool:
#         trie = dict()
#         for w in wordDict:
#             ptr = trie
#             for c in w:
#                 if c not in ptr:
#                     ptr[c] = dict()
#                 ptr = ptr[c]
#             ptr[''] = True
#         dp = [None] * len(s)
#         for i in range(len(s) - 1, -1, -1):
#             ptr = trie
#             for j in range(i, len(s)):
#                 if '' in ptr and dp[j] is True:
#                     dp[i] = True
#                     break
#                 elif s[j] not in ptr:
#                     dp[i] = False
#                     break
#                 else:
#                     ptr = ptr[s[j]]
#             if dp[i] is None:
#                 dp[i] = '' in ptr
#         return dp[0]
